rs/replicated_state/src/canister_snapshots.rs

```diff
@@ -188,6 +188,15 @@ impl CanisterSnapshots {
             snapshot_ids: _,
         } = self;
     }
+
+    /// Returns the amount of memory taken by all canister snapshots on
+    /// this subnet.
+    pub(crate) fn memory_taken(&self) -> NumBytes {
+        self.snapshots
+            .values()
+            .map(|snapshot| snapshot.size())
+            .sum::<NumBytes>()
+    }
```

Comment:
I took a quick look at the `CanisterSnapshots` struct definition, and while its fields are `pub(crate)`, none of them are actually accessed directly from outside of the `CanisterSnapshots` implementation. I would personally make use of that to maintain a precomputed memory usage and thus ensure `O(1)` time for this query.

As a side-note, this may also be a good time to drop the `pub(crate)` qualifier from the `CanisterSnapshots` fields before the very nice to have de facto encapsulation is lost.

Reply:
Good points, thanks! I updated the code to maintain a `memory_usage` field and added a test as well to ensure it's being updated correctly (as well as a debug assert when calling `memory_taken` that checks the precomputed value against the dumb "iterate over all and count").